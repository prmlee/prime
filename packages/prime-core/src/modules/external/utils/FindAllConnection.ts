import { get } from 'lodash';
import { Brackets } from 'typeorm';
import { EntityConnection } from 'typeorm-cursor-connection';
import { DocumentTransformer } from '../../../utils/DocumentTransformer';

export class FindAllConnection<T> extends EntityConnection<T> {
  public documentTransformer: DocumentTransformer;
  public totalCount: number | null = null;

  public createAppliedQueryBuilder(counter = false) {
    const queryBuilder = this.repository.createQueryBuilder();

    if (this.where) {
      (this.where as any)(queryBuilder, counter);
    }

    if (this.afterSelector) {
      queryBuilder.andWhere(this.afterSelector);
    }

    if (this.beforeSelector) {
      queryBuilder.andWhere(this.beforeSelector);
    }

    return queryBuilder;
  }

  public resolveCursor(item: T): string {
    const key = this.sortOptions.map(({ sort }) => {
      if (sort.indexOf('->') === -1) {
        return item[sort.replace(/^\"/, '').replace(/\"$/, '')];
      }
      const path = sort
        .split('->')
        .slice(1)
        .map(n => n.substr(1, n.length - 2))
        .join('.');
      return get(item, `data.${path}`);
    });

    return Buffer.from(JSON.stringify(key)).toString('base64');
  }

  public async query(): Promise<T[]> {
    const { sortOptions } = this;

    const queryBuilder = this.createAppliedQueryBuilder(true);

    const reverse = typeof this.args.last === 'number';
    const appliedOrderMap: any = {
      ASC: reverse ? 'DESC' : 'ASC',
      DESC: reverse ? 'ASC' : 'DESC',
    };

    queryBuilder.orderBy();

    for (const { sort, order } of sortOptions) {
      queryBuilder.addOrderBy(sort, appliedOrderMap[order]);
    }

    if (this.limit) {
      queryBuilder.limit(this.limit);
    }

    const result = await queryBuilder.getRawAndEntities();

    if (reverse) {
      result.entities.reverse();
    }

    result.raw.forEach((item, i) => {
      (result.entities[i] as any).locales = item.locales;
    });

    return result.entities;
  }

  protected keyToSelector(key: any[], direction: 'after' | 'before'): Brackets {
    const eq = direction === 'after' ? ['>', '<', '>=', '<='] : ['<', '>', '<=', '>='];
    const { sortOptions } = this.options;

    return new Brackets(rootQb => {
      for (let i = 0; i < sortOptions.length; i++) {
        const subKeySetComparison = new Brackets(qb => {
          const subKeySet = key.slice(0, i + 1);

          for (let j = 0; j < subKeySet.length; j++) {
            const { sort, order } = sortOptions[j];
            const cursorKey = subKeySet[j];
            const paramterName = Buffer.from(`${direction}__${sort}`).toString('hex');
            let realSort = sort;

            // allow JSONB sorting
            if (sort.indexOf('->') >= 0) {
              const arrowSort = sort.slice(0).split('->');
              const lastSort = arrowSort.pop();
              realSort = `${arrowSort.join('->')}->>${lastSort}`;
            }

            let equality: string;
            if (j === i) {
              equality = order === 'ASC' ? eq[0] : eq[1];
            } else {
              equality = order === 'ASC' ? eq[2] : eq[3];
            }
            qb.andWhere(`${realSort} ${equality} :${paramterName}`, {
              [paramterName]: cursorKey,
            });
          }
        });
        rootQb.orWhere(subKeySetComparison);
      }
    });
  }
}
