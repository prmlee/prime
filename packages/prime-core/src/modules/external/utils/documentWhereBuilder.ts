import { Brackets, WhereExpression } from 'typeorm';
import { SchemaField } from '../../../entities/SchemaField';

interface Where {
  gt?: any;
  gte?: any;
  lt?: any;
  lte?: any;
  eq?: any;
  in?: any;
  contains?: any;
}

export interface NestedWhere {
  [key: string]: undefined | Where | NestedWhere[];
  OR?: NestedWhere[];
  AND?: NestedWhere[];
}

const operators = {
  gt: '>',
  gte: '>=',
  lt: '<',
  lte: '<=',
  eq: '=',
  in: 'IN',
  contains: 'LIKE',
  not: '!=',
};

const modes = { OR: 'orWhere', AND: 'andWhere' };

export const documentWhereBuilder = (
  tableName: string,
  fields: SchemaField[],
  queryBuilder: WhereExpression,
  where: Where | NestedWhere,
  scope: string[] = [],
  mode: string = 'andWhere'
) => {
  for (const [fieldName, whereOrValue] of Object.entries(where)) {
    if (fieldName === 'OR' || fieldName === 'AND') {
      queryBuilder[mode](
        new Brackets(builder => {
          whereOrValue.forEach(innerWhere =>
            documentWhereBuilder(tableName, fields, builder, innerWhere, scope, modes[fieldName])
          );
          return builder;
        })
      );
    } else if (operators.hasOwnProperty(fieldName)) {
      const innerScope = ['"data"', ...scope.slice(0).map(n => `'${n}'`)];
      const lastScopeItem = innerScope.pop();
      const column = `"${tableName}".${innerScope.join('->')}->>${lastScopeItem}`;
      const operator = operators[fieldName];
      if (!operator) {
        continue;
      }
      let value = whereOrValue;
      if (fieldName === 'contains') {
        value = `%${value}%`;
      }
      const key = Buffer.from(value).toString('hex');
      queryBuilder[mode](`${column} ${operator} :${key}`, { [key]: value });
    } else if (whereOrValue) {
      const field = fields.find(
        targetField =>
          targetField.name === fieldName &&
          targetField.parentFieldId === (scope[scope.length - 1] || null)
      );
      if (field && field.primeField) {
        const nextScope = [...scope, field.id];
        documentWhereBuilder(tableName, fields, queryBuilder, whereOrValue, nextScope, mode);
      }
    }
  }
};
