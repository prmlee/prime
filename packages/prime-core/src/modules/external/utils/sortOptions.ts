import { UserInputError } from 'apollo-server-core';
import { SchemaField } from '../../../entities/SchemaField';

export type Order = 'ASC' | 'DESC';

export interface NestedSort {
  [key: string]: Order | NestedSort;
}

export interface SortOption {
  sort: string;
  order: 'ASC' | 'DESC';
}

export const getSortOptions = (
  tableName: string,
  fields: SchemaField[],
  sort: NestedSort | NestedSort[],
  scope: string[] = [],
  acc: SortOption[] = [],
  orderScopes = new Set()
) => {
  if (Array.isArray(sort)) {
    sort.forEach(item => getSortOptions(tableName, fields, item, scope, acc));
  } else {
    for (const [fieldName, dirOrSort] of Object.entries(sort)) {
      const field = fields.find(
        f => f.name === fieldName && f.parentFieldId === (scope[scope.length - 1] || null)
      );
      if (field && field.primeField) {
        const nextScope = [...scope, field.id];
        const scopeKey = nextScope.map(s => `'${s}'`).join('->');
        if (orderScopes.has(scopeKey)) {
          throw new UserInputError('Cannot sort by same field twice');
        }
        orderScopes.add(scopeKey);
        if (typeof dirOrSort === 'string') {
          acc.push({ sort: `"${tableName}"."data"->${scopeKey}`, order: dirOrSort });
        } else {
          getSortOptions(tableName, fields, dirOrSort, nextScope, acc);
        }
      }
    }
  }
  return acc;
};
