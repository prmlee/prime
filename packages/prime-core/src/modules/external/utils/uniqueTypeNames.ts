const cache = new Set();

export const uniqueTypeName = (name, i = 0) => {
  if (cache.has(name)) {
    return uniqueTypeName(`${name}${i + 1}`, i + 1);
  }
  cache.add(name);
  return name;
};

export const clearTypeNames = () => {
  cache.clear();
};
