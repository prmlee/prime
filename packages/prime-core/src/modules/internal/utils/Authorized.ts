import { AuthenticationError } from 'apollo-server-core';
import { UseMiddleware } from 'type-graphql';
import { Context } from '../../../interfaces/Context';

export function Authorized(ruleFn?: any) {
  return UseMiddleware(
    async ({ args, context, info }: { args: any; context: Context; info: any }, next) => {
      const { session = {} } = info || {};
      if (!context.user && !session.user) {
        throw new AuthenticationError('Must be authenticated');
      }

      if (ruleFn) {
        ruleFn(
          {
            can: (action: string, subject: any, field?: string) => {
              context.ability.throwUnlessCan(action, subject, field);
            },
          },
          args
        );
      }

      return next();
    }
  );
}
