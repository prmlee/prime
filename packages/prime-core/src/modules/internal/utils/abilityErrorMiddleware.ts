import { ForbiddenError as AbilityForbiddenError } from '@casl/ability';
import { ForbiddenError } from 'apollo-server-core';
import { MiddlewareFn } from 'type-graphql';

export const abilityForbiddenMiddleware: MiddlewareFn<any> = async ({ context, info }, next) => {
  try {
    return await next();
  } catch (err) {
    if (err instanceof AbilityForbiddenError) {
      throw new ForbiddenError(err.message);
    }
    throw err;
  }
};
