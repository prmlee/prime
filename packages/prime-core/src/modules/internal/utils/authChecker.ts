import { AuthChecker } from 'type-graphql';

type Context = any;

export const authChecker: AuthChecker<Context> = ({ root, args, context, info }, roles) => {
  if (!context.user) {
    return false;
  }
  return true;
};
