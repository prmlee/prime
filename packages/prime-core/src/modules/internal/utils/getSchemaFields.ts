import { defaultsDeep, get } from 'lodash';
import { getRepository, In } from 'typeorm';
import { Schema, SchemaVariant } from '../../../entities/Schema';
import { SchemaField } from '../../../entities/SchemaField';
import { fields as allFields } from '../../../utils/fields';

export const getSchemaFields = async (schemaId: string, inheritance = true) => {
  const schemaRepo = getRepository(Schema);
  const schemaFieldRepo = getRepository(SchemaField);
  const schemaIds = [schemaId];
  const schema = await schemaRepo.findOne(schemaId);

  if (schema && inheritance) {
    schemaIds.push(...get(schema, 'settings.schemaIds', []));
  }

  const fieldsSource = await schemaFieldRepo.find({
    where: {
      schemaId: In(schemaIds),
    },
    order: { position: 'ASC' },
  });

  const fields = [
    ...fieldsSource.filter(f => f.schemaId === schemaId),
    ...fieldsSource.filter(f => f.schemaId !== schemaId),
  ];

  const defaultGroup = schema && schema.variant === SchemaVariant.Template ? schema.title : 'Main';
  const groups = ((schema && schema.groups) || [defaultGroup]).map(title => ({
    title,
    fields: [],
  }));

  const withOptions = field => {
    const fieldInstance = allFields.find(f => f.type === field.type);
    const defaultOptions = (fieldInstance && fieldInstance.defaultOptions) || {};
    field.options = defaultsDeep(field.options || {}, defaultOptions);
    return field;
  };

  fields.reduce((acc, field, index: number) => {
    if (field.parentFieldId) {
      return acc;
    }

    if (!field.group || field.group === '') {
      field.group = defaultGroup;
    }

    let group = groups.find(g => g && g.title === field.group);

    if (!group) {
      group = { title: field.group, fields: [] };
      groups.push(group);
    }

    if (group.fields) {
      group.fields.push(
        withOptions({
          ...field,
          position: field.position || index,
        })
      );
    }

    return acc;
  }, []);

  fields.forEach(field => {
    if (field.parentFieldId) {
      const parentField = fields.find(f => f.id === field.parentFieldId);
      if (parentField) {
        const group = groups.find(g => g.title === parentField.group);
        if (group && group.fields) {
          const target = group.fields.find(f => f.id === parentField.id);
          if (target) {
            target.fields = target.fields || [];
            target.fields.push(
              withOptions({
                ...field,
                position: field.position || target.fields.length,
              })
            );
          }
        }
      }
    }
  });

  return groups;
};
