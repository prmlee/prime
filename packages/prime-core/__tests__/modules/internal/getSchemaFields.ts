import * as TypeORM from 'typeorm';

(TypeORM as any).getRepository = jest.fn(Model => ({
  async findOne() {
    const m = new Model();
    m.id = 123;
    m.variant = 0;
    m.groups = ['Main'];
    return m;
  },
  async find() {
    const field = new Model();
    field.id = 100;
    field.name = 'f1';
    field.schemaId = 123;
    const field2 = new Model();
    field2.id = 300;
    field2.name = 'f3';
    field2.group = 'Other';
    const field3 = new Model();
    field3.id = 200;
    field3.name = 'f2';
    field3.schemaId = 123;
    field3.parentFieldId = 100;
    field.fields = [field3];
    return [field, field2, field3];
  },
}));

import { getSchemaFields } from '../../../src/modules/internal/utils/getSchemaFields';

describe('getSchemaFields', () => {
  it('should match snapshot', async () => {
    expect(await getSchemaFields('123')).toMatchSnapshot();
  });
});
