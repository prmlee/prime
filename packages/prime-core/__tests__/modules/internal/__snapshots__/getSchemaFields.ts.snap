// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSchemaFields should match snapshot 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "fields": Array [
          SchemaField {
            "id": 200,
            "name": "f2",
            "parentFieldId": 100,
            "schemaId": 123,
          },
          Object {
            "id": 200,
            "name": "f2",
            "options": Object {},
            "parentFieldId": 100,
            "position": 1,
            "schemaId": 123,
          },
        ],
        "group": "Main",
        "id": 100,
        "name": "f1",
        "options": Object {},
        "position": 0,
        "schemaId": 123,
      },
    ],
    "title": "Main",
  },
  Object {
    "fields": Array [
      Object {
        "group": "Other",
        "id": 300,
        "name": "f3",
        "options": Object {},
        "position": 1,
      },
    ],
    "title": "Other",
  },
]
`;
